### PR TITLE
MODINV-941 Inventory cannot process Holdings with virtual fields

### PR DESCRIPTION
**Pull request by JavokhirAbdullayev**
### Purpose
Inventory cannot process Holdings with virtual fields [MODINV-941](https://issues.folio.org/browse/MODINV-941)

### Approach
customize object mapper

### Learning
https://javarevisited.blogspot.com/2018/01/how-to-ignore-unknown-properties-parsing-json-java-jackson.html#axzz8NAd4SgHb

### CODE REVIEW

File: src/test/java/org/folio/inventory/storage/external/ExternalStorageModuleHoldingsRecordCollectionExamples.java [hunk 1]
```diff
@@ -33,12 +33,14 @@ public void shouldMapFromJson() {
     assertEquals(permanentLocationId, holdingsrecord.getPermanentLocationId());
   }
 
-  @Test(expected = JsonMappingException.class)
-  public void shouldNotMapFromJsonAndThrowException() {
+  @Test
+  public void shouldNotMapFromJsonAndNotThrowException() {
```
Comment (KaterynaSenchenko) [inline src/test/java/org/folio/inventory/storage/external/ExternalStorageModuleHoldingsRecordCollectionExamples.java:37]:
```suggestion
  public void shouldMapFromJsonAndNotThrowException() {
```

File: src/test/java/api/items/ItemApiMoveExamples.java [hunk 2]
```diff
@@ -164,12 +163,11 @@ public void cannotMoveItemToHoldingsThatDoesNotMatchSchema() {
 
     final var moveItemsResponse = moveItems(newHoldingsId, item);
 
-    assertThat(moveItemsResponse.getStatusCode(), is(500));
+    assertThat(moveItemsResponse.getStatusCode(), is(200));
```
Comment (KaterynaSenchenko) [inline src/test/java/api/items/ItemApiMoveExamples.java:166]: Logic of the test changed - at least, it should be renamed

File: src/main/java/org/folio/inventory/storage/external/ExternalStorageModuleHoldingsRecordCollection.java [hunk 3]
```diff
@@ -1,22 +1,27 @@
...
-
 class ExternalStorageModuleHoldingsRecordCollection
   extends ExternalStorageModuleCollection<HoldingsRecord>
   implements HoldingsRecordCollection {
 
   private static final Logger LOGGER = LogManager.getLogger(ExternalStorageModuleHoldingsRecordCollection.class);
+  private static final ObjectMapper mapper = ObjectMapperTool.getMapper();
+
+  static {
+    mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
```
Comment (KaterynaSenchenko) [inline src/main/java/org/folio/inventory/storage/external/ExternalStorageModuleHoldingsRecordCollection.java:23]: please add a comment why we do this

**Review by Aliaksandr-Fedasiuk — APPROVED** [review]

File: NEWS.md [hunk 4]
```diff
@@ -7,6 +7,7 @@
 * Upgrade folio-kafka-wrapper to 3.0.0 version ([MODINV-834](https://issues.folio.org/browse/MODINV-834))
 * Add deduplication mechanism for sharing Instance ([MODINV-866](https://issues.folio.org/browse/MODINV-866))
 * Allow to overlay source 'MARC' instances without related MARC record ([MODINV-846](https://issues.folio.org/browse/MODINV-847))
+* Inventory cannot process Holdings with virtual fields ([MODINV-941](https://issues.folio.org/browse/MODINV-941))
```
Comment (KaterynaSenchenko) [inline NEWS.md:10]: this should go under new iteration. v20.1.0 has been released, new iteration v20.2.0-SNAPSHOT

**Review by RomanChernetskyi — APPROVED** [review]

File: src/main/java/org/folio/inventory/storage/external/ExternalStorageModuleHoldingsRecordCollection.java [hunk 5]
```diff
@@ -1,22 +1,28 @@
...
-import io.vertx.core.json.JsonObject;
-
 class ExternalStorageModuleHoldingsRecordCollection
   extends ExternalStorageModuleCollection<HoldingsRecord>
   implements HoldingsRecordCollection {
 
   private static final Logger LOGGER = LogManager.getLogger(ExternalStorageModuleHoldingsRecordCollection.class);
+  private static final ObjectMapper mapper = ObjectMapperTool.getMapper();
+
+  // added to ignore unknown fields
```
Comment (KaterynaSenchenko) [inline src/main/java/org/folio/inventory/storage/external/ExternalStorageModuleHoldingsRecordCollection.java:22]: We override the regular mapping due to a specific reason - response from mod-inventory-storage has changed and cannot be mapped to a HoldingsRecord schema. Please list specific fields that we need to ignore. It is a workaround (we ignoring all unknown properties), and it should be acknowledged here

File: src/test/java/api/items/ItemApiMoveExamples.java [hunk 6]
```diff
@@ -150,7 +149,7 @@ public void cannotMoveToNonExistedHoldingsRecord() {
   }
 
   @Test
-  public void cannotMoveItemToHoldingsThatDoesNotMatchSchema() {
+  public void canMoveItemToHoldingsThatDoesNotMatchSchema() {
```
Comment (KaterynaSenchenko) [inline src/test/java/api/items/ItemApiMoveExamples.java:152]:
```suggestion
  public void canMoveItemToHoldings() {
```

File: src/test/java/org/folio/inventory/storage/external/ExternalStorageModuleHoldingsRecordCollectionExamples.java [hunk 7]
```diff
@@ -33,12 +33,14 @@ public void shouldMapFromJson() {
     assertEquals(permanentLocationId, holdingsrecord.getPermanentLocationId());
   }
 
-  @Test(expected = JsonMappingException.class)
-  public void shouldNotMapFromJsonAndThrowException() {
+  @Test
+  public void shouldMapFromJsonAndNotThrowException() {
```
Comment (KaterynaSenchenko) [inline src/test/java/org/folio/inventory/storage/external/ExternalStorageModuleHoldingsRecordCollectionExamples.java:37]:
```suggestion
  public void shouldMapFromJsonAndIgnoreUnknownProperties() {
```

File: src/test/java/org/folio/inventory/storage/external/ExternalStorageModuleHoldingsRecordCollectionExamples.java [hunk 8]
```diff
@@ -33,12 +33,14 @@ public void shouldMapFromJson() {
...
     JsonObject holdingsRecord = new JsonObject()
-      .put("testField", "testValue");
+      .put("holdingsItems", "testValue")
+      .put("bareHoldingsItems", "testValue");
 
-    storage.mapFromJson(holdingsRecord);
+    var result = storage.mapFromJson(holdingsRecord);
+    assertNotNull(result);
```
Comment (KaterynaSenchenko) [inline src/test/java/org/folio/inventory/storage/external/ExternalStorageModuleHoldingsRecordCollectionExamples.java:43]: maybe add some accepted properties, and verify they were mapped

**Review by KaterynaSenchenko — APPROVED** [review]

**Pull request merged**